### ISSUE TITLE
Support log-on-failure statement

### ISSUE THREAD

**Issue by robinp-tw** [jlink/jqwik#172]
Hedgehog has a nice statement that lets you display a value only in case of test failure. This is useful to not only see the inputs to the failed test, but some intermediate states as well. Would something like that hard to support?

Great tool btw, thanks!

**Comment by jlink** [comment]
Sounds useful. Do you have a pointer to the hedgehog feature description?

**Comment by robinp-tw** [comment]
Sure - see the `annotate` and `footnote` functions at https://hackage.haskell.org/package/hedgehog-1.0.4/docs/Hedgehog.html .

**Comment by jlink** [comment]
@robinp-tw Thnaks!

The functionality of `annotate` is roughly covered by `Reporter`: https://jqwik.net/docs/current/user-guide.html#platform-reporting-with-reporter-object

Would something similar, e.g. named `OnFailureReporter` suite your needs?

**Comment by robinp-tw** [comment]
That sounds good! Do I get it correctly the current reporter would log for each test case executed, which might be excessive if we only want the logs in case of failure?

**Comment by jlink** [comment]
@robinp-tw Exactly. The current reporter will always report, regardless of sucess or failure.

**Comment by jlink** [comment]
I've prioritized the feature for the next release (1.5.5).
What do you think about a property to restrict reporting to failure only.

e.g.

```java
@Property
@Report(mode=ON_FAILURE)
void myProperty(Reporter reporter) {
    reporter.publishValue("key", "value");
}
```

Or might it be necessary to have onfailure reports and standard reports in a single property?

**Comment by robinp-tw** [comment]
I only used failure-reporting actively when wanted to debug an error, so I expect it to be turned on only temporarily. Sounds good for the use-case, thank you.

**Comment by jlink** [comment]
My current idea is to offer a `publishValueOnFailure` method:

```
@Property
@Report(mode=ON_FAILURE)
void myProperty(Reporter reporter) {
    reporter.publishValueOnFailure("key1", "value1");
    reporter.publishValueOnFailure("key2", "value2");
}
```

My question: Where would you expect those values to show up? One possibility I see is as part of the test report, e.g.

```
                              |-------------------jqwik-------------------
tries = 2                     | # of calls to property
checks = 2                    | # of not rejected calls
generation = RANDOMIZED       | parameters are randomly generated
after-failure = PREVIOUS_SEED | use the previous seed
when-fixed-seed = ALLOW       | fixing the random seed is allowed
edge-cases#mode = MIXIN       | edge cases are mixed in
edge-cases#total = 9          | # of all combined edge cases
edge-cases#tried = 1          | # of edge cases tried in current run
seed = 4454399202554984829    | random seed to reproduce generated values
key1 | value1
key2 | value2
```

The problem is: For which test runs should the key/value pairs reported:
- For the original failing try?
- For the shrunk try?
- For both?
- If both, where to display the values so that they can easily be attributed to the actual parameters used in this try?

**Comment by jlink** [comment]
Here's the current state of implementation (which can be tested using 1.5.5-SNAPSHOT):

```java
@EnableFootnotes
public class Experiments {

	@BeforeTry
	void beforeTry(Footnotes footnotes) {
		footnotes.addFootnote("footnote before try");
	}

	@Property(tries = 10)
	void test(@ForAll int anInt, Footnotes footnotes) {
		footnotes.addFootnote("number: " + anInt);
		Assertions.assertThat(anInt).isLessThanOrEqualTo(99);
	}

	@AfterTry
	void afterTry(Footnotes footnotes) {
		footnotes.addFootnote("footnote after try");
	}
}
```

Footnotes will be reported for the shrunk sample and the original sample:

```
Shrunk Sample (4 steps)
-----------------------
  anInt: 100
  footnotes: net.jqwik.api.footnotes.FootnotesHook$$Lambda$352/251210093@2d3379b4

  #1 footnote before try
  #2 number: 100
  #3 footnote after try

Original Sample
---------------
  anInt: 153544
  footnotes: net.jqwik.api.footnotes.FootnotesHook$$Lambda$352/251210093@2b71e916

  #1 footnote before try
  #2 number: 153544
  #3 footnote after try
```

Feedback about better API and better reporting welcome!

Instead of `EnableFootnotes` the hook could be made global, but then it would take a performance hit on each invocation of any property method.

**Comment by jlink** [comment]
An alternative is to offer a static method `Footnotes.addFootnote(..)`.

**Comment by jlink** [comment]
Available in 1.5.5-SNAPSHOT.
